### PR TITLE
[#446] rework build jenkinsfile

### LINKED ISSUE #446: Rework Build Process to avoid any host Jenkins dependencies

### PR DESCRIPTION
**Pull request by MaksymBilenko**
This PR fixes #446

### CODE REVIEW

**Comment by MaksymBilenko** [comment]
Added commit for fix #457

**Comment by MaksymBilenko** [comment]
ok to test

**Comment by MaksymBilenko** [comment]
ok to test

**Comment by MaksymBilenko** [comment]
ok to test

**Comment by MaksymBilenko** [comment]
ok to test

**Comment by MaksymBilenko** [comment]
ok to test

**Comment by alexander-semenets** [comment]
we should update version.py before python package build so it has actual metadata. So we should update pipeline with Build&Upload stage which will update version.py, build legion python packages and upload them to pypi within container.

**Comment by MaksymBilenko** [comment]
ok to test

**Comment by MaksymBilenko** [comment]
ok to test

**Comment by MaksymBilenko** [comment]
test this please

**Comment by MaksymBilenko** [comment]
test this please

**Comment by MaksymBilenko** [comment]
test this please

File: pipeline.Dockerfile [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+RUN curl -fsSLO https://download.docker.com/linux/static/stable/x86_64/docker-${DOCKERVERSION}.tgz \
+  && tar xzvf docker-${DOCKERVERSION}.tgz --strip 1 \
+                 -C /usr/local/bin docker/docker \
+  && rm docker-${DOCKERVERSION}.tgz
+
+RUN apt-get update && apt-get install -y software-properties-common \
+	&& echo 'deb http://ppa.launchpad.net/ansible/ansible/ubuntu trusty main' >> /etc/apt/sources.list \
+	&& apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 93C4A3FD7BB9C367 \
+	&& apt-get update \
+	&& curl -sL https://deb.nodesource.com/setup_8.x | bash - \
```
Comment (alexander-semenets) [inline pipeline.Dockerfile]: please remove this as was agreed

**Comment by MaksymBilenko** [comment]
test this please

File: pipeline.Dockerfile (hunk 1 above)
Comment (MaksymBilenko) [reply pipeline.Dockerfile]: Docker and nodejs removed,
Regarding ansible repo - it's ok

**Comment by MaksymBilenko** [comment]
test this please

**Comment by MaksymBilenko** [comment]
test this please

**Comment by MaksymBilenko** [comment]
test this please

File: Jenkinsfile [hunk 2]
```diff
@@ -6,402 +6,494 @@ class Globals {
...
-                    sh """
-                    cp legion/legion/version.py legion_airflow/legion_airflow/version.py
-                    cd legion_airflow
-                    ../.venv/bin/pip install -r requirements/base.txt
-                    ../.venv/bin/pip install -r requirements/test.txt
-                    ../.venv/bin/python3.6 setup.py sdist 
-                    ../.venv/bin/python3.6 setup.py bdist_wheel
-                    ../.venv/bin/python3.6 setup.py develop
-                    """
-                }, 'Build docs': {
```
[the remark this answers is not in the document]
Comment (MaksymBilenko) [reply Jenkinsfile]: Fixed

**Comment by MaksymBilenko** [comment]
test this please

**Comment by alexander-semenets** [comment]
LGTM

File: pipeline.Dockerfile [hunk 3]
```diff
@@ -0,0 +1,34 @@
+FROM python:3.6
+
+ENV DOCKERVERSION=18.03.1-ce
+
+RUN apt-get update && apt-get install -y software-properties-common \
```
Comment (kirillmakhonin) [inline pipeline.Dockerfile]: Please pin versions

File: pipeline.Dockerfile [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	&& apt-get install -y build-essential libssl-dev libffi-dev zlib1g-dev libjpeg-dev git \
+	&& apt-get clean all
+
+RUN pip install Sphinx==1.8.0 sphinx_rtd_theme==0.4.1 sphinx-autobuild==0.7.1 recommonmark==0.4.0 pydocstyle==2.1.1 twine==1.11.0 boto==2.49.0 boto3==1.7.28 ansible==2.6.4 awscli
```
Comment (kirillmakhonin) [inline pipeline.Dockerfile]: Please pin versions and move to requirements file

File: pipeline.Dockerfile [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+ADD legion /src/legion
+RUN cd /src/legion \
+	&& pip install -r requirements/base.txt \
+	&& pip install -r requirements/test.txt \
+	&& python setup.py sdist \
```
Comment (kirillmakhonin) [inline pipeline.Dockerfile]: sdist, bdist_wheel can be removed

File: Jenkinsfile [hunk 6]
```diff
@@ -6,402 +6,492 @@ class Globals {
...
-                ../.venv/bin/pip install -r requirements/base.txt
-                ../.venv/bin/pip install -r requirements/test.txt
-                cd -
-                cd legion_test
-                ../.venv/bin/python3.6 setup.py develop
-                '''
+        }
+        stage('Build Agent Docker Image') {
+            steps {
+                sh "docker build -t legion-docker-agent:${env.BUILD_NUMBER} -f pipeline.Dockerfile ."
```
Comment (dsuslov) [inline Jenkinsfile]: @MaksymBilenko, please use version that is generated or defined for whole build and for all resulting artifacts/containers

Comment (MaksymBilenko) [reply Jenkinsfile]: @dsuslov  Version cannot be generated in the same way before legion build

File: pipeline.Dockerfile (hunk 5 above)
Comment (MaksymBilenko) [reply pipeline.Dockerfile]: done

File: Jenkinsfile [hunk 7]
```diff
@@ -6,402 +6,492 @@ class Globals {
...
+                            }
+                        }
+                    }
+                }
+            }
+        }
+	}
+    post { 
+        always { 
+            deleteDir()
```
Comment (dsuslov) [inline Jenkinsfile]: @MaksymBilenko add notifyBuild() here

File: Jenkinsfile (hunk 6 above)
Comment (dsuslov) [reply Jenkinsfile]: @MaksymBilenko ok. version generation approach should be simplified and refactored. While it is not done, please make legion-docker-agent version more unique. For instance you can add short commit hash.

**Comment by alexander-semenets** [comment]
bot, test all

File: Jenkinsfile [hunk 8]
```diff
@@ -4,413 +4,490 @@ class Globals {
...
+
+                    Globals.dockerCacheArg = (params.EnableDockerCache) ? '' : '--no-cache'
+
+                    Globals.dockerLabels = "--label git_revision=${Globals.rootCommit} --label build_id=${env.BUILD_NUMBER} --label build_user=${env.BUILD_USER} --label build_date=${buildDate}"
+                    println(Globals.dockerLabels)
+                    
+                    /// Define build version
+                    if (params.StableRelease) {
+                        if (params.ReleaseVersion){
+                            Globals.buildVersion = sh returnStdout: true, script: "python3.6 tools/update_version_id --build-version=${params.ReleaseVersion} legion/legion/version.py ${env.BUILD_NUMBER} ${env.BUILD_USER}"
```
Comment (dsuslov) [inline Jenkinsfile]: @alexander-semenets why tools/update_version_id is executed twice here ?

**Comment by dsuslov** [comment]
@alexander-semenets please, check that in case of exceptions and non-zero return code in any stage/step of pipeline notifyBuild() in post{always { } } section is executed

**Comment by alexander-semenets** [comment]
post tested at #624 and #622 runs

**Comment by alexander-semenets** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test app

**Comment by alexander-semenets** [comment]
bot, test app

**Comment by alexander-semenets** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test app

**Comment by alexander-semenets** [comment]
bot, test app

**Comment by alexander-semenets** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test all

**Comment by alexander-semenets** [comment]
bot, test app

**Comment by alexander-semenets** [comment]
bot, test app

**Comment by alexander-semenets** [comment]
bot, test app

**Comment by alexander-semenets** [comment]
bot, test app

**Pull request merged**